### DISCUSSION TITLE
High level explaination

### DISCUSSION THREAD

**Discussion by imechura** [grblHAL/core#420] [General]
Can anyone share a description of how GRBLHal converts gcode into step/pulse signals. Im hoping for discussion level description with references to specific locations in the source code. My ultimate goal is to try to convert some unsupported driver types in the future but admittidly, it will be a very long time before I understand enough to undertake that effort.  I know I should just sit down and figure it all out but my C/C++ is a hobby level so I am not as good at traversing the source as I might be in languages that I have experience with.

**Comment by terjeio** [comment]
There is an overview [here](https://awesome.tech/grbl-demystified/).

> My ultimate goal is to try to convert some unsupported driver types in the future

I do not understand what exactly you mean by this.
If you want to write a driver for a new processor then you do not need to understand the details of the core functionality. The easiest option will be to start with an existing driver, comment out all gpio/peripheral accesses and get serial communication up and running. From there add in support for gpio/peripherals piece by piece. Add in native USB and/or networking support as the last step if required.

**Comment by imechura** [comment]
I understand what you are saying. I suspect you didnt understand me because I was not completley forthcoming with my true intentions. 

As I aluded to before, I t want to get GRBLHal working with TMCM-035 stepper drivers configured to drive the motion via digital SPI waveforms.

 But even moreso than that I want to know/understand how GRBLHal works internally because I use GRBL and GRBLHal on CNC machines on a daily basis and it feels kind of weird that I'm depending on this software so much when I do not actually understand how it works. To be clear these are not 3D printer or desktop CNC routers. Its a full sized plasma table with a 120 AMP power source that cuts 1.25" steel plate and Wells Index Knee mill CNC conversion. The work I do on these machines has, at times, contributed to a significant portion of my income so I am personally, professionally and commercially invested into your software.  
 
I've been a software engineer by trade for 22 years so I am confident that I can figure it out through self-study of the source code however I must assume that an overview of the system's working principlals is likeley to accelerate the pace. My professional experience has been on Java, Spring, JavaScript, Python and I have done a handful of personal projects in C/C++ using Arduino and AVRc. Before I made the upgrade to GRBLHal I developed a stand-alone torch height controller using Arduino for my GRBL1.1 based CNC plasma table. That is the extent of my embeded experience.
